packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -740,6 +756,95 @@ export function runConnectedCallback(vm: VM) {
     }
 }
 
+function setupContext(vm: VM) {
+    const contextKeys = getContextKeys();
+
+    if (isUndefined(contextKeys)) {
+        return;
+    }
+
+    const { connectContext } = contextKeys;
+    const { component, renderer } = vm;
+    const enumerableKeys = keys(getPrototypeOf(component));
+    const contextfulFieldsOrProps = ArrayFilter.call(
+        enumerableKeys,
+        (propName) => (component as any)[propName]?.[connectContext]
+    );
+
+    if (contextfulFieldsOrProps.length === 0) {
+        return;
+    }
+
+    let isProvidingContext = false;
+    const providedContextVarieties: ContextVarieties = new Map();
+
+    const contextRuntimeAdapter: ContextRuntimeAdapter<LightningElement> = {
+        component,
+        provideContext<T extends object>(
+            contextVariety: T,
+            providedContextSignal: Signal<unknown>
+        ): void {
+            if (!isProvidingContext) {
+                isProvidingContext = true;
+                renderer.registerContextProvider(component, ContextEventName, (payload) => {
+                    return payload.setNewContext(providedContextVarieties);
+                });
+            }
+            if (providedContextVarieties.has(contextVariety)) {
+                logErrorOnce(
+                    'Multiple contexts of the same variety were provided. Only the first context will be used.'
+                );
+                return;
+            }
+            providedContextVarieties.set(contextVariety, providedContextSignal);
+        },
+        consumeContext<T extends object>(
+            contextVariety: T,
+            contextProvidedCallback: ContextProvidedCallback
+        ): void {
+            renderer.registerContextConsumer(component, ContextEventName, {
+                setNewContext: (contextVarieties: ContextVarieties) => {
```

Comment:
The return type is `boolean`, which we discussed and I'm on board with. But it might be worth using an aliased type like `type ShouldContinueBubbling = boolean` and setting that explicitly as the return type here for readability. You sort of cover it in the comment below, but this would make it even clearer.